Create FieldMapper GeoAlgorithm

arnaud-morvan opened the pull request:
Objective of this new QGIS GeoAlgorithm is to provide a user friendly tool to refactor layer fields.

It takes a standard ParameterVector as input.

From there, the user can :
- Change some field's names or type.
- Add or delete some fields.
- Reorder fields.
- Calculate new fields based on expressions.
- Load the fields from another loaded layer.

The result is write through a standard OutputVector parameter.
## In modeler

Fully fonctionnal, but save fields mapping as a string (not json list with indentation) because of ModelerAlgorithme.resolveValue behavior with lists.
## In batch mode

Does not display custom widget in batch dialog, seems to be difficult to provide the complex multiline widget in the batch dialog for the field mapping, but can be used with the same syntax as in console mode, exemple:
[{'expression': u'CODE_DEPT', 'length': 2, 'type': 10, 'name': u'CODE_DEPT', 'precision': 0}, ...]
## In console mode

``` python
import processing
processing.runandload("qgis:refactorfields",
                      "DEPARTEMENT",
                      [{'name': u'CODE_DEPT',
                        'type': 10,
                        'length': 2,
                        'precision': 0,
                        'expression': u'CODE_DEPT'
                       }, {
                        'name': u'NOM_DEPT',
                        'type': 10,
                        'length': 30,
                        'precision': 0,
                        'expression': u'NOM_DEPT'
                      }],
                      None)
```

pcav commented:
@volaya @alexbruy

arnaud-morvan commented:
Is it desired that python code from ui files in processing.ui were added to git and not ignored.

alexbruy commented:
It is not clear for me why we need a layer selector in FieldsMappingPanel? Seems it is not needed as we already defined input layer with separate parameter.

Also I suggest to replace all push buttons: "Add", "Delete", "Up", "Down" and "Reset" with tool buttons with icons and drop horizontal lines. This will keep consistency in UI.

And finally, this algorithm does not work for me if fields contain non-ASCII characters.
```
Traceback (most recent call last):
  File "/home/alex/devel/cpp/qgis/build/output/python/plugins/processing/core/GeoAlgorithm.py", line 234, in execute
    self.processAlgorithm(progress)
  File "/home/alex/devel/cpp/qgis/build/output/python/plugins/processing/algs/qgis/FieldsMapper.py", line 83, in processAlgorithm
    expression.evalErrorString()))
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-1: ordinal not in range(128)
```
Same expression works fine from field calculator

arnaud-morvan commented:
@alexbruy Thanks for the review.

The "Load layer fields" feature can be very useful in modeler, to get fields from an existing layer (because we can't get from input layer). Maybe we can hide it in standard dialog.

I've added button icons, and removed separation lines, like requested.

I've fixed evalErrorString encoding problem.

Is it desired that python code from ui files in processing.ui (and algs/qgis/ui) were added to git and not ignored.
It seems that they are produced at make time, so I have'nt commited fieldsmapper related ones.

What about enabling translation in processing plugin (inserting tr() calls for names, groups and parameters), don't you work on this at Essen hackfest ?

alexbruy commented:
I'm a bit busy now, will try to take a look at Friday.

Regarding translation, you can use tr() calls for parameter names and messages

alexbruy commented:
It seems that algorithm does not work in batch mode, at least I can not find any way to set "Fields mapping" parameter from batch interface.

Still not sure that "Load layer fields" functionality is necessary as it depends on layers that loaded in QGIS and will not work if algorithm runs with empty project. Maybe better make parameter available in modeler, so one can create as many mappings as necessary and add them to algorithm.

@volaya what do you think?

arnaud-morvan commented:
@alexbruy @volaya 
Is there a reason why names and groups are not passed through translation ?
Ou customers would like groups and names translated to french.
Can I make a pull request for this ?

arnaud-morvan commented:
Rebased on master

alexbruy commented:
Right now group and algorithm names used in various parts of Processing for generating commandline names and accessing help files. I'm working on this already.

arnaud-morvan commented:
@alexbruy @volaya 
Maybe this could be merge now.

Same as #1759, I'm not completly aware of parameter method getValueAsCommandLineParameter, in which context is it used ?

Parameter FieldsMapping does not display in batch dialog, but it seems difficult to provide a friendly user interface in the batch dialog for the fields mapping.

volaya commented:
I like the idea of the algorithm, but, as Alex said, I am sceptic about adding a new parameter type, specially such a specific one. I don't understand the point of this one, specially in the modeler, since it is not connected to other parameters (it cannot be filed with the output of another algorithm, so it is just a hardcoded parameter). In this case, i would recommend creating a custom modeler dialog for the algorithm, instead of adding a new parameter that doesn't seem to be very useful in other context and will add complexity to the modeler code itself.

What do you think? Maybe we can start adding the algorithm with a string parameter instead of you custom p arameter (so you have the algorithm available although not with a custom interface and just a text input box as widget), and then work on the custom interface.

arnaud-morvan commented:
My first strategy was to add a custom dialog for this algorithm, extending from standard dialog. But this was very tricky and hard to reproduce in the modeler because the two dialogs are'nt constructed the same way (use of ParametersPanel for the first and not for the second). It was largely dependant on the precise base dialogs structure, and I was afraid that a very small change in standard or modeler dialog would break the custom dialogs.

Adding a new "specific" parameter seems to be a very modular and clear way to get similar UI in the two dialogs, no chances that a small change in standard or modeler dialog break this geoalgorithm UI.

With this specific, parameter, fields definitions are easily readable in modeler files (because of json indentation) and easy to modify by hand.

For now, the developpement is completly finished and fixed, and it's a lot of work to change this now, impossible for the feature freeze.

Maybe this "specific" parameter seems too specific, but it can be easily reused, and it don't bring a lot of complexity in standard and modeler dialog, it adds some ifs, depending on parameter type. Maybe, later, I would add another algorithm to copy/paste features from one layer to another, which suppose a similar fields mapping.

arnaud-morvan commented:
After some reflexion, I agree with you, I will propose a new, less instrusive, version today.

arnaud-morvan committed:
Create FieldMapper GeoAlgorithm

arnaud-morvan commented:
Thanks again for review.
This new version of FieldsMapper (completely squashed to avoid conflicts) works exactly the same way but is less intrusive with processing core and gui packages.
Hoping that it could be merged before freeze.

arnaud-morvan commented:
@volaya I've rebased on master to avoid conflicts
tested with batch, modeler and console, all works fine.
